namerd/iface/interpreter-thrift/src/main/scala/io/buoyant/namerd/iface/ThriftNamerInterface.scala

```diff
@@ -100,14 +100,16 @@ object ThriftNamerInterface {
     protected[this] def nextStamp(): Stamp
 
     protected[this] final def update(v: Try[T]): Unit = {
-      val obs = Observation(nextStamp(), v)
-      val promise = synchronized {
-        val previous = pending
-        current = Some(obs)
-        pending = new Promise[Observation]
-        previous
+      if (!current.exists(_.value == v)) {
+        val obs = Observation(nextStamp(), v)
+        val promise = synchronized {
+          val previous = pending
+          current = Some(obs)
```

Comment:
I see what you mean.  I don't immediately see how this could go wrong, but that doesn't mean that it can't.  To be safe I've expanded the synchronized block to the whole method.  In reality there should only be one updater so I don't think we'll really need to worry about contention.  Better thread-safe than sorry.